File: Part 1 - Java/MatchingCost.java
//Name: Vidulash Rajaratnam
//SNum: 8190398

import java.io.*;
import java.util.ArrayList;
import java.util.List;

//Class responsible to calculate the matching cost through 2 methods
public class MatchingCost {
    //Method that creates csv file and returns cost matrix for two lists of faces using euclidean distance algorithm
    public int[][] usingEuclideanDistance(List<Face> faces) throws IOException {
        List<Face> frame1 = new ArrayList<>();
        List<Face> frame2= new ArrayList<>();
        for (Face f:faces){
            if (f.getFrame() == 1){
                frame1.add(f); }
            else{
                frame2.add(f); } }

        File file = new File("matching_costs_euclidean.csv");
        FileWriter outputfile = new FileWriter(file);
        BufferedWriter bw = new BufferedWriter(outputfile);
        PrintWriter pw = new PrintWriter(bw);

        String[] header = new String[frame2.size()+1];
        header[0] = ",";
        for (int i = 1;i<=frame1.size();i++){
            if (i==frame1.size()){
                header[i] = frame1.get(i-1).getName();}
            else{
                header[i] = frame1.get(i-1).getName()+",";
            }
        }
        pw.println(getStringOutput(header));

        for (int i = 0;i<frame1.size();i++){
            //Initialize size
            String[] line = new String[frame2.size()+1];
            //Add frame name first with comma
            line[0] = frame2.get(i).getName()+",";
            //For loop that adds names from first frame along with matching distances
            for (int x =0;x<frame2.size();x++){
                if (x==frame2.size()-1){
                    line[x+1] = Integer.toString(Euclidean(frame1.get(i),frame2.get(x)));}
                else{
                    line[x+1] = Integer.toString(Euclidean(frame1.get(i),frame2.get(x)))+","; }
            }
            //Writes to file
            pw.println(getStringOutput(line));
        }
        pw.flush();
        pw.close();

        //For loop to create matrix of distances
        int[][] matrix = new int[frame1.size()][frame2.size()];
        for (int i = 0;i<frame1.size();i++){
            int[] line = new int[frame1.size()];
            for (int x = 0;x<frame2.size();x++){
                line[x] = Euclidean(frame1.get(i),frame2.get(x));
            }
            matrix[i] = line;
        }

        //Returns matching matrix in array of array of integers
        return matrix;
    }

    //Function that calculates euclidean distance from two faces
    private int Euclidean(Face faceone, Face facetwo){
        //Getting center coordinates for both faces
        double[] c_faceone = new double[]{(faceone.getUpX()+(faceone.getUpX()+faceone.getWidthX()))/2,(faceone.getUpY()+(faceone.getUpY()+faceone.getHeightY()))/2};
        double[] c_facetwo = new double[]{(facetwo.getUpX()+(facetwo.getUpX()+facetwo.getWidthX()))/2,(facetwo.getUpY()+(facetwo.getUpY()+facetwo.getHeightY()))/2};
        double x = Math.pow((c_facetwo[0]-c_faceone[0]),2);
        double y = Math.pow((c_facetwo[1]-c_faceone[1]),2);
        //Actually calculate distance
        int distance = (int) Math.round(Math.sqrt(x+y));
        return distance;
    }

    //Method that creates csv file and returns cost matrix for two lists of faces using area of box
    public int[][] usingAreaofBox(List<Face> faces) throws IOException {
        List<Face> frame1 = new ArrayList<>();
        List<Face> frame2= new ArrayList<>();
        for (Face f:faces){
            if (f.getFrame() == 1){
                frame1.add(f); }
            else{
                frame2.add(f); } }

        File file = new File("matching_costs_area.csv");
        FileWriter outputfile = new FileWriter(file);
        BufferedWriter bw = new BufferedWriter(outputfile);
        PrintWriter pw = new PrintWriter(bw);

        String[] header = new String[frame2.size()+1];
        header[0] = ",";
        for (int i = 1;i<=frame1.size();i++){
            if (i==frame1.size()){
                header[i] = frame1.get(i-1).getName();}
            else{
                header[i] = frame1.get(i-1).getName()+",";
            }
        }
        pw.println(getStringOutput(header));

        //
        for (int i = 0;i<frame1.size();i++){
            //Initialize size
            String[] line = new String[frame2.size()+1];
            //Add frame name first with comma
            line[0] = frame2.get(i).getName()+",";
            //For loop that adds names from first frame along with matching areas
            for (int x =0;x<frame2.size();x++){
                if (x==frame2.size()-1){
                    line[x+1] = Integer.toString(Math.abs(getFrameArea(frame1.get(x))-getFrameArea(frame2.get(i))));}
                else{
                    line[x+1] = Integer.toString(Math.abs(getFrameArea(frame1.get(x))-getFrameArea(frame2.get(i))))+","; }
            }
            //Writes to file
            pw.println(getStringOutput(line));
        }
        pw.flush();
        pw.close();

        //For loop to create matrix of distances
        int[][] matrix = new int[frame1.size()][frame2.size()];
        for (int i = 0;i<frame1.size();i++){
            int[] line = new int[frame1.size()];
            for (int x = 0;x<frame2.size();x++){
                line[x] = (Math.abs(getFrameArea(frame1.get(x))-getFrameArea(frame2.get(i))));
            }
            matrix[i] = line;
        }

        //Returns matching matrix in array of array of integers
        return matrix;
    }

    //Method that returns area of a frame
    private int getFrameArea(Face frame){
        int area = frame.getWidthX()*frame.getHeightY();
        return area;
    }

    //Function that outputs string representation of string array, for proper format to add to csv file
    private String getStringOutput(String[] result){
        String output = "";
        for (int i =0;i<result.length;i++){
            output += result[i];
        }
        return output;
    }
}

File: Part 2 - Golang/MatchingCost/MatchingCost.go
//NAME: Vidulash Rajaratnam
//SNUM: 8190398

package MatchingCost

import (
	"fmt"
	"math"
	"strconv"
)

//Function that prints out [][]int slice in a clear matrix format
func PrintMatrix(matrix [][]int){
	fmt.Print("Current Matrix:\n")
	for i:= 0;i<len(matrix);i++{
		fmt.Print("[ ")
		for n:= 0;n<len(matrix[i]);n++{
			fmt.Print(matrix[i][n]," ")
		}
		fmt.Print("] \n")
	}
	fmt.Print("\n")
}

//Function to calculate euclidean distance and form resulting slice of integers
func MatchingCostEuclidean(frame1 [][]string, frame2 [][]string)(result [][]int){
	for i:= 0;i<len(frame1);i++{
		var line []int
		for n:= 0;n<len(frame2);n++{
			value:= Euclidean(frame1[i],frame2[n])
			line = append(line,value)
		}
		result= append(result,line)

	}
	fmt.Print("----------EUCLIDEAN METHOD----------\n")
	PrintMatrix(result)
	return result
}

//Function that calculates the euclidean distance between two frames
func Euclidean(frameone []string, frametwo []string) (result int){
	oneh1,_ := strconv.ParseFloat(frameone[1],64)
	oneh2,_ := strconv.ParseFloat(frameone[3],64)
	oneh := (oneh1+oneh2+oneh1)/2 //145.5

	onew1,_ := strconv.ParseFloat(frameone[2],64)
	onew2,_ := strconv.ParseFloat(frameone[4],64)
	onew := (onew1+onew2+onew1)/2 //145.5

	twoh1,_ := strconv.ParseFloat(frametwo[1],64)
	twoh2,_ := strconv.ParseFloat(frametwo[3],64)
	twoh := (twoh1+twoh2+twoh1)/2 //145.5

	twow1,_ := strconv.ParseFloat(frametwo[2],64)
	twow2,_ := strconv.ParseFloat(frametwo[4],64)
	twow := (twow1+twow2+twow1)/2 //145.5

	result = int(math.Round(math.Pow(math.Pow(float64(twoh-oneh),2) + math.Pow(float64(twow-onew),2),0.5)))
	return result
}

//Function to calculate box area and form resulting slice of integers
func MatchingCostBoxArea(frame1 [][]string, frame2 [][]string)(result [][]int){
	for i:= 0;i<len(frame1);i++{
		var line []int
		for n:= 0;n<len(frame2);n++{
			value:= Boxarea(frame1[n],frame2[i])
			line = append(line,value)
		}
		result= append(result,line)

	}
	fmt.Print("----------AREA METHOD----------\n")
	PrintMatrix(result)
	return result
}

//Function that calculates the box area of two frames and returns their difference
func Boxarea(frameone []string, frametwo []string) (result int){
	oneh,_ := strconv.ParseFloat(frameone[3],64)
	onew,_ := strconv.ParseFloat(frameone[4],64)
	twoh,_ := strconv.ParseFloat(frametwo[3],64)
	twow,_ := strconv.ParseFloat(frametwo[4],64)
	f1area := oneh * onew
	f2area := twoh * twow
	result = int(math.Round(math.Abs(f1area - f2area)))
	return result
}
